src/graphapi/graphapi_tensor.cpp

```diff
@@ -0,0 +1,312 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <miopen/graphapi/graphapi_tensor.hpp>
+#include <miopen/errors.hpp>
+
+#include <algorithm>
+
+namespace miopen {
+
+namespace graphapi {
+
+TensorBuilder& TensorBuilder::setDataType(miopenDataType_t dataType) &
+{
+    mTensor.setDataType(dataType);
+    mDataTypeSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setDim(const std::vector<int64_t>& dimensions) &
+{
+    if(dimensions.empty() ||
+       std::any_of(dimensions.cbegin(), dimensions.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setDimensions(dimensions);
+    mDimensionsSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setDim(std::vector<int64_t>&& dimensions) &
+{
+    if(dimensions.empty() ||
+       std::any_of(dimensions.cbegin(), dimensions.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setDimensions(std::move(dimensions));
+    mDimensionsSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setStride(const std::vector<int64_t>& strides) &
+{
+    if(strides.empty() ||
+       std::any_of(strides.cbegin(), strides.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setStrides(strides);
+    mStridesSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setStride(std::vector<int64_t>&& strides) &
+{
+    if(strides.empty() ||
+       std::any_of(strides.cbegin(), strides.cend(), [](int64_t val) { return val <= 0; }))
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mTensor.setStrides(std::move(strides));
+    mStridesSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setId(int64_t id) &
+{
+    mTensor.setId(id);
+    mUniqueIdSet = true;
+    return *this;
+}
+
+TensorBuilder& TensorBuilder::setVirtual(bool isVirtual) &
+{
+    mTensor.setVirtual(isVirtual);
+    return *this;
+}
+
+Tensor TensorBuilder::build() const&
+{
+    if(!mUniqueIdSet || !mDataTypeSet || !mDimensionsSet || !mStridesSet ||
+       mTensor.getDimensions().size() != mTensor.getStrides().size())
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    return mTensor;
+}
+
+Tensor TensorBuilder::build() &&
+{
+    if(!mUniqueIdSet || !mDataTypeSet || !mDimensionsSet || !mStridesSet ||
+       mTensor.getDimensions().size() != mTensor.getStrides().size())
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    return std::move(mTensor);
```

Comment:
[Quality] @CAHEK7 @amberhassaan @scerzh Please notice that std:move changes the state of mTensor to the moved-from state, which makes it unusable. The above means that after `build()` the objection should not be used anymore. Do we guarantee/enforce that? Maybe it is worth adding `IsBuilt` or `IsUnusable` flag and check it where appropriate?

Reply:
There are 2 overloads of `build()`. The declaration of this one is enough here.
```C++
Tensor TensorBuilder::build() &&
```
it is making the whole builder object unusable: you should not access the builder after it, including the proposed call of `IsUnusable()`; cppcheck or similar analyzers should trigger errors if you do. For the same reason STL doesn't have such methods for its containers.

Reply:
@atamazov : the `isFinalized()` method is what you want from `isUsable()`. Once the object has been finalized, the builder is no longer there. We have thought of replacing the builder with a `std::optional<builder-type>` and marking the std::optional false after `Finalize()`. And we have discussed keeping a pointer to `builder` and making it null. I think all are more or less equivalent solutions.

Reply:
@amberhassaan @scerzh @CAHEK7 Yeah, the idea is like to set the flag and throw an exception in all methods of the class if it is set (except dtor).

Reply:
OK, I see the value in the idea.

We don't need an additional flag. We already have flags causing `build()&` and `build()&&` to throw if something required wasn't set. We can reset the state of the builder to initial.